executor/src/witgen/jit/single_step_processor.rs

```diff
@@ -164,9 +228,55 @@ mod test {
         let err = generate_single_step(input, "M").err().unwrap();
         assert_eq!(
             err.to_string(),
-            "Unable to derive algorithm to compute values for witness columns in the next row for the following columns:\n\
+            "Unable to derive algorithm to compute values for witness columns in the next row and\n\
+            unable to branch on a variable. The following columns are still missing:\n\
             M::Y\n\
-            and 0 identities are missing."
+            The following identities have not been fully processed:\n"
+        );
+    }
+
+    #[test]
+    fn branching() {
+        let input = "
+    namespace VM(256);
+        let A: col;
+        let B: col;
+        let instr_add: col;
+        let instr_mul: col;
+        let pc: col;
+
+        col fixed LINE = [0, 1] + [2]*;
+        col fixed INSTR_ADD = [0, 1] +  [0]*;
+        col fixed INSTR_MUL = [1, 0] + [1]*;
+
+        pc' = pc + 1;
+        [ pc, instr_add, instr_mul ] in [ LINE, INSTR_ADD, INSTR_MUL ];
+
+        instr_add * (A' - (A + B)) + instr_mul * (A' - A * B)  + (1 - instr_add - instr_mul) * (A' - A) = 0;
+        B' = B;
+        ";
+        let code = generate_single_step(input, "Main").unwrap();
+        assert_eq!(
+            format_code(&code),
+            "\
+VM::pc[1] = (VM::pc[0] + 1);
+machine_call(1, [Known(VM::pc[1]), Unknown(ret(1, 1, 1)), Unknown(ret(1, 1, 2))]);
+VM::instr_add[1] = ret(1, 1, 1);
+VM::instr_mul[1] = ret(1, 1, 2);
+VM::B[1] = VM::B[0];
+if (VM::instr_add[0] == 1) {
+    if (VM::instr_mul[0] == 1) {
+        VM::A[1] = -((-(VM::A[0] + VM::B[0]) + -(VM::A[0] * VM::B[0])) + VM::A[0]);
```

Comment:
Interesting. It doesn't know that instruction flags are exclusive. Does that mean that for $n$ instructions it generates $2^n$ branches? (For the Keccak RISC-V example, $n = 36$.)